README.markdown

[TrustCommerce](http://www.trustcommerce.com) is a payment gateway providing credit card
processing and recurring / subscription billing services.

This library provides a simple interface to create, edit, delete, and query subscriptions
using TrustCommerce.

## Background ##

TrustCommerce's recurring / subscription billing solution is implemented through a service
called [Citadel](http://www.trustcommerce.com/citadel.php).  A Citadel-enabled account is
required to use the subscription-based features implemented by this library.

### Citadel Basics ###

* Citadel stores customer profiles which can include credit card information and billing frequency.
* Citadel will automatically bill customers on their respective schedules.
* Citadel identifies each customer by a Billing ID (six-character alphanumeric string).
* A customer's profile, credit card, and billing frequency can be modified using the Billing ID.

## Installation ##

The simple way:

    $ sudo gem install trustcommerce

Directly from repository:

    $ svn co svn://rubyforge.org/var/svn/trustcommerce/trunk trustcommerce

It is highly recommended to download and install the
[TCLink ruby extension](http://www.trustcommerce.com/tclink.php).
This extension provides failover capability and enhanced security features.
If this library is not installed, standard POST over SSL will be used.

## Configuration ##

When you signup for a TrustCommerce account you are issued a custid and a password.
These are your credentials when using the TrustCommerce API.

    TrustCommerce.custid   = '123456'
    TrustCommerce.password = 'topsecret'

    # optional - sets Vault password for use in query() calls
    TrustCommerce.vault_password = 'supersecure'

The password that TrustCommerce issues never changes or expires when used through the TCLink
extension.  However if you choose to use SSL over HTTP instead (the fallback option if the TCLink
library is not installed), be aware that you need to set the password to your Vault password.
Likewise, if your application uses the query() method you must set the vault_password.
The reason is that TrustCommerce currently routes these query() calls through the vault 
and therefore your password must be set accordingly.  To make matters more complicated, 
TrustCommerce currently forces you to change the Vault password every 90 days.


## Examples ##

### Creating a subscription ###

    # Bill Jennifer $12.00 monthly
    response = TrustCommerce::Subscription.create(
      :cc     => '4111111111111111',
      :exp    => '0412',
      :name   => 'Jennifer Smith',
      :amount => 1200,
      :cycle  => '1m'
    )
    
    if response['status'] == 'approved'
      puts "Subscription created with Billing ID: #{response['billingid']}"
    else
      puts "An error occurred: #{response['error']}"
    end


### Update a subscription ###

    # Update subscription to use new credit card
    response = TrustCommerce::Subscription.update(
      :billingid => 'ABC123',
      :cc        => '5411111111111115', 
      :exp       => '0412'
    )
    
    if response['status'] == 'accepted'
      puts 'Subscription updated.'
    else
      puts "An error occurred: #{response['error']}"
    end


### Delete a subscription ###

    # Delete subscription
    response = TrustCommerce::Subscription.delete(
      :billingid => 'ABC123'
    )
    
    if response['status'] == 'accepted'
      puts 'Subscription removed from active use.'
    else
      puts 'An error occurred.'
    end


### Query a subscription ###

    # Get all sale transactions for a subscription in CSV format
    response = TrustCommerce::Subscription.query(
      :querytype => 'transaction',
      :action    => 'sale',
      :billingid => 'ABC123'
    )


### Process a one-time charge ###

    # Process one-time sale against existing subscription
    response = TrustCommerce::Subscription.charge(
      :billingid => 'ABC123',
      :amount    => 1995
    )


### Credit a transaction ###

    # Process one-time credit against existing transaction
    response = TrustCommerce::Subscription.credit(
      :transid => '001-0000111101',
      :amount  => 1995
    )


## Running the tests ##

The following special environment variables must be set up prior to running tests:

    $ export TC_USERNAME=123456
    $ export TC_PASSWORD=password
    $ export TC_VAULT_PASSWORD=password

Run tests via rake:

    $ rake test

Run tests via ruby:

    $ ruby test/trustcommerce_test.rb



CHANGELOG
trunk:

0.5.3:

Fixed small typo in statuses: defined should be decline [Ara Vartanian <avartanian@postful-inc.com>]

0.5.2:

Add TrustcommerceSyncAgent as example code on how to sync transactions with the Vault.

0.5.1:

Return responses as a Result class (smarter Hash) to allow stuff like: response.accepted? [Carl Lerche <carl.lerche@verizon.net>]

0.5.0:

- Initial public release

test/test_helper.rb
require 'test/unit'
require 'rubygems'
require File.dirname(__FILE__) + '/../lib/trustcommerce'

begin
  require 'mocha'
rescue LoadError
  puts "Mocha required to run tests. `gem install mocha` and try again."
  exit 1
end

TCLINK_LIB_INSTALLED = TrustCommerce.tclink?

def via_tclink_and_https
  # test via TCLink
  if !TCLINK_LIB_INSTALLED
    puts 'TCLink library not installed - skipping tests via TCLink'
  else
    TrustCommerce.stubs(:tclink?).returns(true)
    yield
  end
  
  # test via https
  TrustCommerce.stubs(:tclink?).returns(false)
  yield
end

def create_subscription!(name)
  response = TrustCommerce::Subscription.create(
    :cc       => CARDS[:visa][:cc], 
    :exp      => CARDS[:visa][:exp],
    :address1 => CARDS[:visa][:address],
    :zip      => CARDS[:visa][:zip],
    :avs      => 'y',      
    :name     => name,
  	:amount   => 1200,
  	:cycle    => '1m',
    :demo     => 'y'
  )
  assert_equal 'approved', response[:status]    
  assert response.keys.include?(:billingid)
  response[:billingid]
end

# --- [ TrustCommerce test data ] ---
# reference: https://vault.trustcommerce.com/downloads/TCDevGuide.html#testdata
CARDS = {
  :visa       => { :cc      => '4111111111111111', 
                   :exp     => '0412', 
                   :cvv     => 123,
                   :address => '123 Test St.', 
                   :city    => 'Somewhere', 
                   :state   => 'CA', 
                   :zip     => 90001 },
  :mastercard => { :cc      => '5411111111111115', 
                   :exp     => '0412', 
                   :cvv     => 777,
                   :address => '4000 Main St.', 
                   :city    => 'Anytown', 
                   :state   => 'MA', 
                   :zip     => 85001 },
  :amex       => { :cc      => '341111111111111',  
                   :exp     => '0412', 
                   :cvv     => 4000,
                   :address => '12 Colorado Blvd.', 
                   :city    => 'Elsewhere', 
                   :state   => 'IL', 
                   :zip     => 54321 }
}

test/trustcommerce_test.rb
require File.join(File.dirname(__FILE__), 'test_helper')

# The following special environment variables must be set up prior to running tests:
# 
#     $ export TC_USERNAME=123456
#     $ export TC_PASSWORD=password
#     $ export TC_VAULT_PASSWORD=password
#
# Run tests via rake:
#
#     $ rake test
#
# Run tests via ruby:
#
#     $ ruby test/trustcommerce_test.rb
class TrustCommerceSubscriptionTest < Test::Unit::TestCase

  def setup
    if ENV['TC_USERNAME'].nil? || ENV['TC_PASSWORD'].nil?
      puts 'TC_USERNAME and TC_PASSWORD must be set.'
      puts 'Usage: TC_USERNAME=username TC_PASSWORD=password TC_VAULT_PASSWORD=password ruby test/trustcommerce_test.rb'
      exit 1
    else
      TrustCommerce.custid         = ENV['TC_USERNAME']
      TrustCommerce.password       = ENV['TC_PASSWORD']
      TrustCommerce.vault_password = ENV['TC_VAULT_PASSWORD'] if ENV['TC_VAULT_PASSWORD']
    end
  end

  def test_subscription_create
    via_tclink_and_https do
      response = TrustCommerce::Subscription.create(
        :cc       => CARDS[:visa][:cc], 
        :exp      => CARDS[:visa][:exp], 
        :address1 => CARDS[:visa][:address],
        :zip      => CARDS[:visa][:zip],
        :avs      => 'y',
        :name     => 'Jennifer Smith - create() test',
      	:amount   => 1200,
      	:cycle    => '1m',
        :demo     => 'y'
      )
      assert_equal TrustCommerce::Result, response.class
      assert_not_nil response[:billingid]
      assert response.keys.include?(:transid)
      assert_equal 'approved', response[:status]  
      assert response.approved?
    end
  end
  
  def test_subscription_update
    via_tclink_and_https do
      billing_id = create_subscription!('update() test')
      response = TrustCommerce::Subscription.update(
        :billingid => billing_id,
        :cc        => CARDS[:mastercard][:cc], 
        :exp       => CARDS[:mastercard][:exp],
        :address1  => CARDS[:mastercard][:address],
        :zip       => CARDS[:mastercard][:zip],
        :avs       => 'y'
      )
      assert_equal 'accepted', response[:status]
      assert response.accepted?
    end
  end
  
  def test_subscription_delete
    via_tclink_and_https do
      billing_id = create_subscription!('delete() test')
      response = TrustCommerce::Subscription.delete(
        :billingid => billing_id
      )
      assert response.keys.include?(:transid)
      assert_equal 'accepted', response[:status]
      assert response.accepted?
    end
  end
  
  def test_subscription_charge_and_credit
    via_tclink_and_https do    
      billing_id = create_subscription!('charge and credit() test')
      
      # charge
      charge_response = TrustCommerce::Subscription.charge(
        :billingid  => billing_id,
        :amount     => 1995,
        :demo       => 'y'
      )
      assert charge_response.keys.include?(:transid)
      assert_equal 'approved', charge_response[:status]
      assert charge_response.approved?
    
      # credit
      credit_response = TrustCommerce::Subscription.credit(
        :transid  => charge_response[:transid],
        :amount   => 995,
        :demo     => 'y'
      )
      assert credit_response.keys.include?(:transid)
      assert_equal 'accepted', credit_response[:status]
      assert credit_response.accepted?
    end
  end
  
  def test_subscription_query
    puts "\n"
    puts "---------------------------------------------------------------------------"
    puts "IMPORTANT: This query test will likely take between 1 and 2 minutes!"
    puts "Make sure TC_VAULT_PASSWORD is set if it differs from your TCLink password."
    puts "---------------------------------------------------------------------------"
    
    # create subscription
    billing_id = create_subscription!('query() test')

    # query for charges
    options = { :querytype => 'transaction', :action => 'sale', :billingid => billing_id }
    while (query_response = TrustCommerce::Subscription.query(options))
      if query_response.body =~ /error/i
        fail(query_response.body)
        break
      elsif query_response.body.split("\n").size < 2
        puts 'Transaction has not yet showed up... will try again in 15 seconds.'
        sleep(15)
      else
        puts 'Transaction found.'
        
        # setup index hash
        field_names = query_response.body.split("\n")[0].split(',')
        date_line_1 = query_response.body.split("\n")[1].split(',')
        indexes = field_names.inject({}) {|h, field| h[field.to_sym] = field_names.index(field); h }

        # check transaction data
        assert_equal '1111',            date_line_1[indexes[:cc]]
        assert_equal '1200',            date_line_1[indexes[:amount]]
        assert_equal 'query() test',    date_line_1[indexes[:name]]
        break
      end
    end
  end
  
  # test private helpers
  def test_stringify_hash
    assert_equal ({ 'a' => '1', 'b' => '2' }),  TrustCommerce.stringify_hash(:a => '1', :b => '2')
    assert_equal ({ 'a' => '1', 'b' => '2' }),  TrustCommerce.stringify_hash(:a => 1, :b => 2)
    assert_equal ({ 'a' => '2' }),              TrustCommerce.stringify_hash(:a => 1, :a => 2)
  end

  def test_symbolize_hash
    assert_equal ({ :a => '1', :b => '2' }),  TrustCommerce.symbolize_hash('a' => '1', 'b' => '2')
    assert_equal ({ :a => '1', :b => '2' }),  TrustCommerce.symbolize_hash(:a => 1, 'b' => 2)
    assert_equal 1, TrustCommerce.symbolize_hash(:a => 1, 'a' => 2).size
  end
  
end


lib/trustcommerce_sync_agent.rb
#
# This agent will pull down recent transactions from the TrustCommerce vault using the
# following schema: 
#
#   create_table :subscriptions do |t|
#     t.column :account_id,                :integer,   :null => false
#     t.column :created_on,                :datetime,  :null => false
#     t.column :billing_id,                :string
#     t.column :length,                    :integer,   :null => false
#     t.column :cents,                     :integer,   :null => false
#     t.column :billing_full_name,         :string
#     t.column :billing_address,           :string
#     t.column :billing_zip_code,          :string
#     t.column :billing_country,           :string
#     t.column :billing_card_type,         :string
#     t.column :billing_credit_card,       :string
#     t.column :billing_expiration_date,   :datetime   
#   end
#   
#   add_index :subscriptions, :account_id
#   
#   create_table :subscription_transactions do |t|
#     t.column :subscription_id,          :integer,   :null => false
#     t.column :account_id,               :integer,   :null => false
#     t.column :transaction_date,         :datetime,  :null => false
#     t.column :transaction_id,           :string,    :null => false
#     t.column :transaction_type,         :string,    :null => false
#     t.column :amount,                   :decimal,   :null => false
#     t.column :card_number,              :string,    :null => false
#     t.column :card_type,                :string,    :null => false
#     t.column :cardholder_name,          :string,    :null => false     
#   end
#   
#   add_index :subscription_transactions, :subscription_id
#   add_index :subscription_transactions, :account_id
#
#   Install as cron job or test it out on the console:
#
# $ ./script/runner -e production "TrustcommerceSyncAgent.sync('3h')"
#
class TrustcommerceSyncAgent
  
  # time_ago examples: 
  #   30m => 30 minutes
  #   1h  => 1 hour
  #   3d  => 3 days
  def self.sync(time_ago = nil)
    time = case time_ago
      when /^(\d+)m$/ then Time.now.utc - $1.to_i.minute
      when /^(\d+)h$/ then Time.now.utc - $1.to_i.hour
      when /^(\d+)d$/ then Time.now.utc - $1.to_i.day
      else Time.now.utc - 1.hour
    end
    # --- [ find recently created paying subscriptions ] ---
    subscriptions = Subscription.find(:all, :conditions => ['cents > 0 AND created_on > ?', time])
    subscriptions.each do |subscription|
      sync_subscription(subscription)
    end
  end

  private  

    def self.sync_subscription(subscription)
        
      # --- [ get TC data ] ---
      tc = TrustCommerce::Subscription.query(
        :querytype  => 'transaction',
        :billingid  => subscription.billing_id
      )      
      return if !tc.kind_of? Net::HTTPOK

      # --- [ create index by CSV header ]
      field_names = tc.body.split("\n")[0].split(',')
      indexes = field_names.inject({}) {|h, field| h[field.to_sym] = field_names.index(field); h }

      # --- [ build transaction array ] ---
      transactions = tc.body.split("\n")
      transactions.shift # get rid of header

      transactions.each do |line|
        transaction = line.split(',')

        #log_transaction(indexes, transaction)
      
        subscription_transaction = subscription.transactions.find_by_transaction_id(transaction[indexes[:transid]])
        if subscription_transaction.nil?
          SubscriptionTransaction.create(
            :subscription_id    => subscription.id,
            :account_id         => subscription.account_id,
            :transaction_date   => convert_date(transaction[indexes[:trans_date]]),
            :transaction_id     => transaction[indexes[:transid]],
            :transaction_type   => transaction[indexes[:action_name]],
            :amount             => transaction[indexes[:bank_amount]],
            :card_number        => transaction[indexes[:cc]],
            :card_type          => convert_card_type(transaction[indexes[:media_name]]),
            :cardholder_name    => transaction[indexes[:name]]
          )
        end      
      end
    
    end

    # TC returns mm-dd-yyyy HH:mm:ss
    def self.convert_date(str)
      date = str.split(' ')[0].split('-')
      Time.local(date[2], date[0], date[1])
    end
  
    # TC returns VISA-D, MC-D, AMEX-D
    def self.convert_card_type(str)
      case str
        when /VISA/i  then 'Visa'
        when /MC/i    then 'MasterCard'
        when /AMEX/i  then 'American Express'
        else str
      end
    end
    
    # --- [ helpful for debugging ] ---
    def self.log_transaction(indexes, transaction)
      puts '---------- transaction --------------------'
      indexes.each{|k,v| puts "#{k} => #{transaction[indexes[k]]}" if !transaction[indexes[k]].blank? }
      puts '-------------------------------------------'
    end
    
end

lib/trustcommerce.rb
require 'net/http'
require 'net/https'
require 'uri'

# [TrustCommerce](http://www.trustcommerce.com) is a payment gateway providing credit card
# processing and recurring / subscription billing services.
#
# This library provides a simple interface to create, edit, delete, and query subscriptions
# using TrustCommerce.
#
# ## Background ##
#
# TrustCommerce's recurring / subscription billing solution is implemented through a service
# called [Citadel](http://www.trustcommerce.com/citadel.php).  A Citadel-enabled account is
# required to use the subscription-based features implemented by this library.
#
# ### Citadel Basics ###
#
# * Citadel stores customer profiles which can include credit card information and billing frequency.
# * Citadel will automatically bill customers on their respective schedules.
# * Citadel identifies each customer by a Billing ID (six-character alphanumeric string).
# * A customer's profile, credit card, and billing frequency can be modified using the Billing ID.
#
# ## Installation ##
#
# The simple way:
#
#     $ sudo gem install trustcommerce
#
# Directly from repository:
#
#     $ svn co svn://rubyforge.org/var/svn/trustcommerce/trunk trustcommerce
#
# It is highly recommended to download and install the
# [TCLink ruby extension](http://www.trustcommerce.com/tclink.php).
# This extension provides failover capability and enhanced security features.
# If this library is not installed, standard POST over SSL will be used.
#
# ## Configuration ##
#
# When you signup for a TrustCommerce account you are issued a custid and a password.
# These are your credentials when using the TrustCommerce API.
#
#     TrustCommerce.custid   = '123456'
#     TrustCommerce.password = 'topsecret'
#
#     # optional - sets Vault password for use in query() calls
#     TrustCommerce.vault_password = 'supersecure'
#
# The password that TrustCommerce issues never changes or expires when used through the TCLink
# extension.  However if you choose to use SSL over HTTP instead (the fallback option if the TCLink
# library is not installed), be aware that you need to set the password to your Vault password.
# Likewise, if your application uses the query() method you must set the vault_password.
# The reason is that TrustCommerce currently routes these query() calls through the vault 
# and therefore your password must be set accordingly.  To make matters more complicated, 
# TrustCommerce currently forces you to change the Vault password every 90 days.
class TrustCommerce

  class << self
    attr_accessor :custid
    attr_accessor :password
    attr_accessor :vault_password
    
    # Returns Vault password.
    def vault_password
      @vault_password || password
    end    
  end
  self.custid   = 'TestMerchant'
  self.password = 'password'
  
  # Settings for standard POST over SSL 
  # Only used if TCLink library is not installed
  API_SETTINGS = { 
    :domain => 'vault.trustcommerce.com',
    :query_path => '/query/', 
    :trans_path => '/trans/', 
    :port => 443  
  }
  
  class Subscription

    #     # Bill Jennifer $12.00 monthly
    #     response = TrustCommerce::Subscription.create(
    #       :cc     => '4111111111111111',
    #       :exp    => '0412',
    #       :name   => 'Jennifer Smith',
    #       :amount => 1200,
    #       :cycle  => '1m'
    #     )
    #     
    #     if response['status'] == 'approved'
    #       puts "Subscription created with Billing ID: #{response['billingid']}"
    #     else
    #       puts "An error occurred: #{response['error']}"
    #     end
    def self.create(options)
      return TrustCommerce.send_request(options.merge(:action => 'store'))
    end
    
    #     # Update subscription to use new credit card
    #     response = TrustCommerce::Subscription.update(
    #       :billingid => 'ABC123',
    #       :cc        => '5411111111111115', 
    #       :exp       => '0412'
    #     )
    #     
    #     if response['status'] == 'accepted'
    #       puts 'Subscription updated.'
    #     else
    #       puts "An error occurred: #{response['error']}"
    #     end
    def self.update(options)
      return TrustCommerce.send_request(options.merge(:action => 'store'))
    end    
    
    #     # Delete subscription
    #     response = TrustCommerce::Subscription.delete(
    #       :billingid => 'ABC123'
    #     )
    #     
    #     if response['status'] == 'accepted'
    #       puts 'Subscription removed from active use.'
    #     else
    #       puts 'An error occurred.'
    #     end
    def self.delete(options)
      return TrustCommerce.send_request(options.merge(:action => 'unstore'))
    end
    
    #     # Process one-time sale against existing subscription
    #     response = TrustCommerce::Subscription.charge(
    #       :billingid => 'ABC123',
    #       :amount    => 1995
    #     )
    def self.charge(options)
      return TrustCommerce.send_request(options.merge(:action => 'sale'))
    end
    
    #     # Process one-time credit against existing transaction
    #     response = TrustCommerce::Subscription.credit(
    #       :transid => '001-0000111101',
    #       :amount  => 1995
    #     )
    def self.credit(options)
      return TrustCommerce.send_request(options.merge(:action => 'credit'))
    end
    
    #     # Get all sale transactions for a subscription in CSV format
    #     response = TrustCommerce::Subscription.query(
    #       :querytype => 'transaction',
    #       :action    => 'sale',
    #       :billingid => 'ABC123'
    #     )
    def self.query(options)
      return TrustCommerce.send_query(options)
    end    
    
  end
  
  class Result < Hash
    
    def initialize(constructor = {})
      if constructor.is_a?(Hash)
        super()
        update(constructor)
      else
        super(constructor)
      end
    end
    
    %w(approved accepted decline baddata error).each do |status|
      define_method("#{status}?") do
        self[:status] == status
      end
    end
    
  end
  
  # It is highly recommended to download and install the
  # [TCLink ruby extension](http://www.trustcommerce.com/tclink.php).
  # This extension provides failover capability and enhanced security features.
  # If this library is not installed, standard POST over SSL will be used.
  def self.tclink?
    begin
      require 'tclink'
      true
    rescue LoadError
      false
    end
  end
  
  private
  
    def self.stringify_hash(hash)
      hash.inject({}) { |h,(k,v)| h[k.to_s] = v.to_s; h }
    end

    def self.symbolize_hash(hash)
      hash.inject({}) { |h,(k,v)| h[k.to_sym] = v.to_s; h }
    end
    
    def self.send_request(options)
      options[:custid]   = self.custid
      options[:password] = self.password
      options.update(:demo => 'y') if ENV['RAILS_ENV'] != 'production'
      parameters = stringify_hash(options)
      if tclink? # use TCLink extension if installed
        return Result.new(symbolize_hash(TCLink.send(parameters)))
      else # TCLink library not installed - use https post
        parameters[:password] = self.vault_password.to_s
        response = send_https_request(API_SETTINGS[:trans_path], parameters)
        
        # parse response
        results = Result.new
        response.body.split("\n").each do |line|
          k, v = line.split('=')
          results[k.to_sym] = v
        end
        results
      end
    end
    
    def self.send_query(options)
      options[:custid]   = self.custid
      options[:password] = self.vault_password.to_s
      response = send_https_request(API_SETTINGS[:query_path], stringify_hash(options))
    end
    
    def self.send_https_request(path, parameters)
      http = Net::HTTP.new(API_SETTINGS[:domain], API_SETTINGS[:port])
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE # avoid ssl cert warning
      request = Net::HTTP::Post.new(path)
      request.form_data = parameters
      http.request(request)
    end

end